File: adsa/Day6/maxprofit.py
# class Solution:
#     def maxProfit(self, prices: list[int]) -> int:

#         n= len(prices)
#         profit = 0
#         for i in range(n):
#             for j in range(i+1,n):
#                 if prices[j]- prices[i] > profit:
#                     profit = max(profit,prices[j]-prices[i])
        
#         return profit
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        min_price = float('inf')
        max_profit = 0

        for price in prices:
                min_price = min(min_price, price)
                max_profit = max(max_profit, price - min_price)

        return max_profit
prices = list(map(int, input().split()))
sol = Solution().maxProfit(prices)
print(sol)

File: adsa/Day29(dp2)/maxprofit.cpp
#include <bits/stdc++.h> 
#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        int mini=prices[0];
        int maxprofit=0;
        for(int i=1;i<n;i++){
            int cost = prices[i]-mini;
            maxprofit=max(maxprofit,cost);
            mini = min(mini,prices[i]);
        }
        return maxprofit;
        
    }
};
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<int> prices(n);
        for (int i = 0; i < n; i++) {
            cin >> prices[i];
        }
        Solution obj;
        cout << obj.maxProfit(prices) << endl;
    }
    return 0;
}